### PR TITLE
Implementation of some tricks to avoid looped blind clicking

### PR DESCRIPTION
**Pull request by stackhanovets**
Connects to #276

### CODE REVIEW

File: kcauto.sikuli/combat.sikuli/combat.py [hunk 1]
```diff
@@ -181,10 +181,8 @@ def _select_combat_map(self):
 
         if self.lbas:
             # resupply and delay sortie time if LBAS fails fatigue check
-            lbas_check_fatigue = (
-                'CheckFatigue' in self.config.combat['misc_options'])
-            pass_lbas_check, delay_time = (
-                self.lbas.resupply_groups(lbas_check_fatigue))
+            lbas_check_fatigue = 'CheckFatigue' in self.config.combat['misc_options']
+            pass_lbas_check, delay_time = self.lbas.resupply_groups(lbas_check_fatigue)
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/combat.py]: These two lines now fail linting (PEP-8, line length). Please revert the changes.

File: kcauto.sikuli/combat.sikuli/combat.py [hunk 2]
```diff
@@ -404,14 +402,11 @@ def _run_combat_logic(self):
...
 
-                combat_result = self._run_loop_during_battle()
-
-                # resolve night battle
-                if combat_result == 'night_battle':
-                    if self._select_night_battle(self._resolve_night_battle()):
-                        self._run_loop_during_battle()
-
-                self.regions['lower_right_corner'].wait('next.png', 30)
+                while not self.regions['lower_right'].exists('next.png', 1 + Globals.SLEEP_MODIFIER):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/combat.py]: `Util.kc_sleep(1)` exists as a convenience method for things like `1 + Globals.SLEEP_MODIFIER`, with additional variability. Please use it instead. Also, this line, as-is, fails linting (line length).

File: kcauto.sikuli/combat.sikuli/combat.py [hunk 3]
```diff
@@ -404,14 +402,11 @@ def _run_combat_logic(self):
...
+                    # TODO: implement into Util class the recovery call
+                    # if any of recovery assets exists within kc_region
+                    if self.kc_region.exists('combat_nb_fight.png', 1 + Globals.SLEEP_MODIFIER):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/combat.py]: Ditto as comment for `L405`.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 4]
```diff
@@ -5,7 +5,9 @@
 
 
 class LBAS(object):
-    def __init__(self, config, regions, map):
+    _lbas_group_modes_tuple = ("defense", "retreat", "rest", "sortie", "standby")
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: The preferred naming for static variables is all caps, so something like `LBAS_GROUP_MODES` would be more appropriate.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 5]
```diff
@@ -5,7 +5,9 @@
...
+
+    def __init__(self, config, regions, _map):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Was the `map` parameter renamed due to conflict with the internal `map`? If so, lets rename it to `map_data` instead of `_map`.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 6]
```diff
@@ -31,29 +33,27 @@ def assign_groups(self):
         """Method for assigning sortied LBAS groups to their respective nodes
         on the sortie map.
         """
-        self.kc_region.wait('lbas_panel_ready.png', 10)
-        Util.kc_sleep(2)
+        while not self.regions['upper'].exists("lbas_panel_ready.png", 1 + Globals.SLEEP_MODIFIER):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Is it really worth introducing another `while` loop here? Does it actually take more than 10 seconds to call up this panel?

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 7]
```diff
@@ -31,29 +33,27 @@ def assign_groups(self):
...
                 'lbas_group_1_nodes', 'lbas_group_2_nodes',
                 'lbas_group_3_nodes'):
             if not self.config.combat[lbas_group_nodes]:
                 # if no lbas nodes are specified for this group, skip it
                 continue
-
-            Util.log_msg(
-                "Assigning nodes for LBAS group {}.".format(
-                    lbas_group_nodes[11]))
+            Util.log_msg("Assigning nodes for LBAS group #{}.".format(lbas_group_nodes[11]))
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Linting error.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 8]
```diff
@@ -31,29 +33,27 @@ def assign_groups(self):
...
             nodes = self.config.combat[lbas_group_nodes]
             for node in nodes:
                 node_obj = self.map.nodes[node]
                 lbas_sidebar_pos = 'left'
-                lbas_sidebar = self.kc_region.find('lbas_panel_side.png')
-                if lbas_sidebar.x > 600:
+                lbas_sidebar_match = None
+                while not lbas_sidebar_match:
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Ditto on the additional `while` loop here. Does this actually ever fail?

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 9]
```diff
@@ -31,29 +33,27 @@ def assign_groups(self):
...
+                    lbas_sidebar_match = self.kc_region.exists("lbas_panel_side.png", 1 + Globals.SLEEP_MODIFIER)
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Linting error, use `Util.kc_sleep()`.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 10]
```diff
@@ -31,29 +33,27 @@ def assign_groups(self):
...
+                if lbas_sidebar_match.x > 600:
                     lbas_sidebar_pos = 'right'
-                if ((lbas_sidebar_pos == 'left' and node_obj.coords[0] < 420)
-                        or (lbas_sidebar_pos == 'right'
-                            and node_obj.coords[0] > 780)):
+                if ((lbas_sidebar_pos == 'left' and node_obj.coords[0] < 420) or
+                        (lbas_sidebar_pos == 'right' and node_obj.coords[0] > 780)):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Linting error on this conditional; revert change.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 11]
```diff
@@ -137,7 +137,7 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
         group_fatigue = self._check_fatigue()
         self.print_fatigue_states(group)
         if group_fatigue['high'] or group_fatigue['medium']:
-            # lbas group is fatigued; put it in rest mode
+            Util.log_msg("LBAS group #{} is fatigued, assigning to rest mode.".format(group))
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Linting error.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 12]
```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
...
                 if group_fatigue['medium'] else fatigue['medium'])
         else:
+            Util.log_msg("LBAS group #{} has good morale.".format(group))
             lbas_group_nodes_key = 'lbas_group_{}_nodes'.format(group)
             # put LBAS group into air defense mode if it is active but no nodes
             # are assigned to it
             group_sortie_mode = (
                 'sortie'
                 if self.config.combat[lbas_group_nodes_key] else 'defense')
+            Util.log_msg("Assign LBAS group #{a} to {b} mode.".format(a=group, b=group_sortie_mode))
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Linting error. Let's remove the named fields if they're going to be something as vague as `a` and `b`; I don't think they help in readability at all, and it's only two fields.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 13]
```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
...
+    def _switch_lbas_mode(self, lbas_group_mode_name_target):
         """Switches the lbas group mode to the specified final mode.
 
         Args:
-            final_mode (str): the mode to switch the LBAS group to
+            lbas_group_mode_name_target (str): the mode to switch the LBAS group to
         """
         Util.rejigger_mouse(self.regions, 'top')
+        # Must be relevant to http://kancolle.wikia.com/wiki/Land_Base_Aerial_Support#Options
+        if sum([lbas_group_mode_name_target == i for i in self._lbas_group_modes_tuple]) != 1:
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: This can simply be:
```python
if lbas_group_mode_name_target not in self._lbas_group_modes_tuple:
```

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 14]
```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
...
             # are assigned to it
             group_sortie_mode = (
                 'sortie'
                 if self.config.combat[lbas_group_nodes_key] else 'defense')
+            Util.log_msg("Assign LBAS group #{a} to {b} mode.".format(a=group, b=group_sortie_mode))
             self._switch_lbas_mode(group_sortie_mode)
         return fatigue
 
-    def _switch_lbas_mode(self, final_mode):
+    def _switch_lbas_mode(self, lbas_group_mode_name_target):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Most of the lines in this method go over the line limit and cause linting errors. Please revise.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 15]
```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
...
+        # Must be relevant to http://kancolle.wikia.com/wiki/Land_Base_Aerial_Support#Options
+        if sum([lbas_group_mode_name_target == i for i in self._lbas_group_modes_tuple]) != 1:
+            raise ValueError("No such LBAS mode: \"{}\".".format(lbas_group_mode_name_target))
         while not self.module_regions['lbas_mode_switcher'].exists(
-                'lbas_group_mode_{}.png'.format(final_mode)):
-            Util.click_preset_region(self.regions, 'lbas_mode_switch_button')
-            Util.rejigger_mouse(self.regions, 'top')
-            Util.kc_sleep(1)
-        Util.log_msg("LBAS group switched to {} mode.".format(final_mode))
+                'lbas_group_mode_{}.png'.format(lbas_group_mode_name_target), 1 + Globals.SLEEP_MODIFIER):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: Use `Util.kc_sleep` here

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 16]
```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
...
+            for lbas_group_mode_name_current in self._lbas_group_modes_tuple:
+                lbas_group_mode_pattern = Pattern('lbas_group_mode_{}.png'.format(
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: You don't need to call `Pattern().similar()` if using `Globals.DEFAULT_SIMILARITY` as it is defined to every pattern by default at the start of kcauto.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 17]
```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
...
+                    lbas_group_mode_name_current)).similar(Globals.DEFAULT_SIMILARITY)
+                Util.rejigger_mouse(self.regions, 'top')
+                if self.module_regions['lbas_mode_switcher'].exists(
+                        lbas_group_mode_pattern, 1 + Globals.SLEEP_MODIFIER):
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: You could actually just not define `lbas_group_mode_pattern` on `L173` and just use `.exists('lbas_group_mode_{}.png'.format(lbas_group_mode_name_current))`.

Also, use `Util.kc_sleep` here.

File: kcauto.sikuli/combat.sikuli/lbas.py [hunk 18]
```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
...
+                    print("Current LBAS group mode is \"{}\".".format(lbas_group_mode_name_current))
+                    if lbas_group_mode_name_current == lbas_group_mode_name_target:
```
Comment (mrmin123) [inline kcauto.sikuli/combat.sikuli/lbas.py]: I like where you're going trying to remove the blind-click behavior from LBAS group assignment, but as it stands this is a really slow way to do it. This algorithm loops through every single possible matches after every click, until it actually finds the mode it's trying to find. With the limited size of the set (5) one might think this should be a fast operation, but each search is actually in the realm of significant milliseconds, not nanoseconds.

I think the appropriate approach would be this: reorder the values of the tuple so that they're actually in the order they show in-game. In this method, do a `for` search for which mode the squad is currently in. Then, create a re-ordered list that contains which modes we should expect to see on every click with something like this:
```python
expected_modes = self._lbas_group_modes_tuple[n:] + self._lbas_group_modes_tuple[:n]
```
Where `n` is the index of the matched current state. This will ensure that each subsequent click incurs an O(1) cost.

File: kcauto.sikuli/pvp.sikuli/pvp.py [hunk 19]
```diff
@@ -73,40 +73,25 @@ def run_pvp_logic(self):
...
-            pass
-
-        Util.click_preset_region(self.regions, 'center')
-
-        while not self.regions['home_menu'].exists('home_menu_sortie.png'):
-            # click through post-combat screens until main menu
-            Util.click_preset_region(self.regions, 'shipgirl')
-            Util.kc_sleep(2)
-
+        while not self.regions['lower_right_corner'].exists('next.png', 1 + Globals.SLEEP_MODIFIER):
```
Comment (mrmin123) [inline kcauto.sikuli/pvp.sikuli/pvp.py]: Use `Util.kc_sleep` whenever relevant, and fix linting errors.

File: kcauto.sikuli/util.sikuli/util.py [hunk 20]
```diff
@@ -474,6 +474,7 @@ def click_mouse(cls, region):
         region.mouseDown(Button.LEFT)
         cls.kc_sleep()
         region.mouseUp(Button.LEFT)
+        cls.log_msg("Click")
```
Comment (mrmin123) [inline kcauto.sikuli/util.sikuli/util.py]: If we're adding a console log for the click, let's make it more informative. [Env.getMouseLocation](http://doc.sikuli.org/globals.html#Env.getMouseLocation) returns a Sikuli `Location` object where we can get the `x` and `y` location of the mouse. If we add this information to the console log it'll make things easier to debug.

**Review by mrmin123 — CHANGES_REQUESTED** [review]
Thanks for the PR. I've made specific notes in the diff view.

Major points:
* Use `Util.kc_sleep` instead of `1 + Globals.SLEEP_MODIFIER`
* Adhere to PEP8 linting rule requiring line lengths of < 80
* Additional modifications to `_switch_lbas_mode` method logic
* Change names of some variables
* Surface additional information from `click_mouse`

File: kcauto.sikuli/combat.sikuli/combat.py (hunk 1 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/combat.py]: IntelliJ IDEA issue. Refactored.

File: kcauto.sikuli/combat.sikuli/combat.py (hunk 2 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/combat.py]: They are not equal.

* `Util.kc_sleep(1)` actually blocks the thread explicitly  
* `Region.exists(1)` performs a continuous scan until timeout ends ([1](https://raiman.github.io/SikuliX1/javadocs/org/sikuli/script/Region.html#exists-PSI-), [2](https://github.com/RaiMan/SikuliX1/blob/cbbc8f1aea19a3ff1c6a180aa94ab13232acced0/API/src/main/java/org/sikuli/script/Region.java#L2341)). It contains `RepeatableFind.repeat()` method call and returns `Match` before or `null` after the timeout ends. `Region.wait(Pattern, FOREVER)` also does the same, but I do prefer the `while` loop due to flexibility.

 If you're talking about `DRY`, I can just refactor it into `Util` static method, with randomization like `Util.kc_sleep()` provides, so there is no need to call additional pause if not catbombed. What's your strategy?

File: kcauto.sikuli/combat.sikuli/lbas.py (hunk 4 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/lbas.py]: `_LBAS_GROUP_MODES` since it is private static field.

File: kcauto.sikuli/combat.sikuli/lbas.py (hunk 5 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/lbas.py]: There was no conflict. Just do not use reserved words.

File: kcauto.sikuli/combat.sikuli/lbas.py (hunk 16 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/lbas.py]: Sure. Moreover, any constant may be overrided from elsewhere since python does not provide `final` variable modificators.

File: kcauto.sikuli/combat.sikuli/lbas.py (hunk 17 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/lbas.py]: This is linting issue.

File: kcauto.sikuli/combat.sikuli/lbas.py (hunk 18 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/lbas.py]: Have you tested it already? I'll order the tuple, however, it's not slow to implement `CircularBuffer` class since the target `Region` is just tiny. Moreover, there is also fun related with `threading` or `multiprocessing` implementations. And after this we will put here  `Util.kc_sleep()` call to beware of catbombs from fast clicking. Not a first priority, sorry.

File: kcauto.sikuli/util.sikuli/util.py (hunk 20 above)
Comment (stackhanovets) [reply kcauto.sikuli/util.sikuli/util.py]: Will try.

File: kcauto.sikuli/combat.sikuli/combat.py (hunk 2 above)
Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/combat.py]: Ah, you're right. I had a brain fart and automatically started thinking about `Util.kc_sleep` since that's really only been the place where `Globals.SLEEP_MODIFIER` was used. I see what you're trying to do here. Let me think about this a bit.

File: kcauto.sikuli/combat.sikuli/lbas.py (hunk 18 above)
Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/lbas.py]: I did test this earlier and was getting a noticeable delay on 'standby' since it was the last one in the list, but this was on a pretty slow machine. `Thread`s are an option but I it's a big overkill. My proposal was aiming to avoid a major refactor.

Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/lbas.py]: I didn't fix any of the linting errors but this is my take on the logic:

```python
    def _switch_lbas_mode(self, lbas_group_mode_name_target):
        """Switches the lbas group mode to the specified final mode.

        Args:
            lbas_group_mode_name_target (str): the mode to switch the LBAS group to
        """
        Util.rejigger_mouse(self.regions, 'top')
        # Must be relevant to http://kancolle.wikia.com/wiki/Land_Base_Aerial_Support#Options
        if sum([lbas_group_mode_name_target == i for i in self._lbas_group_modes_tuple]) != 1:
            raise ValueError("No such LBAS mode: \"{}\".".format(lbas_group_mode_name_target))
        for idx, lbas_group_mode_name_current in enumerate(self._lbas_group_modes_tuple):
            if self.module_regions['lbas_mode_switcher'].exists(
                    'lbas_group_mode_{}.png'.format(lbas_group_mode_name_current), 1 + Globals.SLEEP_MODIFIER):
                break
        expected_modes = self._lbas_group_modes_tuple[idx:] + self._lbas_group_modes_tuple[:idx]
        for idx, mode in enumerate(expected_modes):
            print("Current LBAS group mode is \"{}\".".format(lbas_group_mode_name_current))
            if mode == lbas_group_mode_name_target:
                break
            Util.check_and_click(self.module_regions['lbas_mode_switcher'], 'lbas_group_mode_{}.png'.format(mode))
            Util.rejigger_mouse(self.regions, 'top')
            self.module_regions['lbas_mode_switcher'].wait('lbas_group_mode_{}.png'.format(expected_modes[idx + 1]))
        Util.log_msg("LBAS group switched to {} mode.".format(lbas_group_mode_name_target))
```

File: kcauto.sikuli/combat.sikuli/lbas.py (hunk 16 above)
Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/lbas.py]: FYI, since you've been responding to the other comments, there are '10 hidden conversations' right above this one you might have missed.

File: kcauto.sikuli/combat.sikuli/combat.py (hunk 2 above)
Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/combat.py]: The more I think about it I feel that most of the `Globals.SLEEP_MODIFIER` modifiers on these `exists` calls are not necessary. The ones in `while` loops naturally resolve themselves if they're just set to 1second, The ones outside of `while` loops can just be blank, which defaults to 3seconds. I don't think people generally set a sleep modifier above 3 in the first place.

File: kcauto.sikuli/pvp.sikuli/pvp.py [hunk 21]
```diff
@@ -73,40 +73,25 @@ def run_pvp_logic(self):
...
+            if self.kc_region.exists('combat_nb_fight.png', 1 + Globals.SLEEP_MODIFIER):
+                if night_battle:
+                    Util.check_and_click(self.kc_region, 'combat_nb_fight.png')
+                else:
+                    Util.check_and_click(self.kc_region, 'combat_nb_retreat.png')
+                Util.rejigger_mouse(self.regions, 'top')
+
+        while not self.regions['home_menu'].exists('home_menu_sortie.png', 1 + Globals.SLEEP_MODIFIER):
+            if self.regions['lower_right_corner'].exists('next.png', 1 + Globals.SLEEP_MODIFIER):
+                Util.check_and_click(self.regions['lower_right_corner'], 'next.png')
```
Comment (mrmin123) [inline kcauto.sikuli/pvp.sikuli/pvp.py]: I personally think that this is an OK place to use `click_preset_region`...

File: kcauto.sikuli/combat.sikuli/combat.py (hunk 2 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/combat.py]: Yes. `What You See is What You Script.` When we're playing a game, we're not just waiting for buttons - we're looking for them. The implementation makes the code very flexible, I can even set up a countdown timer and/or crash assets detection to call recovery within it. The only point that I see as an obstacle is possible CPU overconsumption. But I'd like to disregard it.
Another point is too fast clicking. In theory `Globals.SLEEP_MODIFIER` would help to set up a balance during too fast matching, otherwise the waiting timeout is already provided.

Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/combat.py]: Are you responding to the wrong comment? This comment isn't about `WYSIWYS` at all.

Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/combat.py]: Perhaps you are misunderstanding how `.exists()` works? `Globals.SLEEP_MODIFIER` would not effect the fastest match here. It only effects the acceptable slowest match, which is moot since the since the `.exists()`s are within a `while` loop and persists until a match is found. See [this page](http://doc.sikuli.org/region.html#Region.exists) for a description of the parameters.

File: kcauto.sikuli/pvp.sikuli/pvp.py (hunk 21 above)
Comment (mrmin123) [reply kcauto.sikuli/pvp.sikuli/pvp.py]: If you are going to persist on moving away from `click_preset_region` here (it doesn't really matter to me), please set up a new `EXPAND` keyword in `kca_globals` so that it clicks the center of the screen or the shipgirl area relative to the button. One of the purpose of `click_preset_region` was to provide a convenient way to define a larger click region for anti-anti-bot measures, and I'd like to not lose that.

File: kcauto.sikuli/combat.sikuli/combat.py (hunk 2 above)
Comment (stackhanovets) [reply kcauto.sikuli/combat.sikuli/combat.py]: I was talking about `Util.kc_sleep` application after matching success to prevent fast clicking. And they are not quite the same - [Region.wait()](https://github.com/RaiMan/SikuliX1/blob/cfb86d3d35fb9cfe63f0ad65f0540677924860a0/API/src/main/java/org/sikuli/script/Region.java#L2231) also contains a while loop, [Region.exists()](https://github.com/RaiMan/SikuliX1/blob/cfb86d3d35fb9cfe63f0ad65f0540677924860a0/API/src/main/java/org/sikuli/script/Region.java#L2360) - not. This doesn't make sense at all, I just like `while region.exists(pattern, timeout)` more than, well...
```python
while not good and counter < treshold:
  try:
    region.wait(pattern, timeout)
    good = True
  except FindFailed:
    counter += 1
    if counter == treshold:
      raise RecoveryCallingException("msg")
```

File: kcauto.sikuli/pvp.sikuli/pvp.py (hunk 21 above)
Comment (stackhanovets) [reply kcauto.sikuli/pvp.sikuli/pvp.py]: Sure, `next.png` is small enough.  I'll implement this later

File: kcauto.sikuli/combat.sikuli/combat.py (hunk 2 above)
Comment (mrmin123) [reply kcauto.sikuli/combat.sikuli/combat.py]: You really need to clarify what you're talking about when you jump from topic to topic. I obviously can't read your mind and your tendency to switch topics randomly is leading to a lot of confusion and needless back and forth.

**Comment by mrmin123** [comment]
@stackhanovets if you're going to make sweeping changes to the basic infrastructure and Utility methods, you should first discuss your proposed changes with me first. I'm not about to do reviews for sweeping architectural changes that affect the entire codebase, changes that were not brought up prior, in a Pull Request.

This PR is very close to entering a state where I will not be comfortable accepting and merging.

**Comment by stackhanovets** [comment]
Requesting recheck after 53ca347

File: kcauto.sikuli/util.sikuli/util.py [hunk 22]
```diff
@@ -432,6 +433,69 @@ def rejigger_mouse(cls, regions, preset):
 
         regions['game'].mouseMove(Location(rand_x, rand_y))
 
+    @staticmethod
```
Comment (mrmin123) [inline kcauto.sikuli/util.sikuli/util.py]: Again, please remove the three new methods you added in `Util` and refactor code in the rest of your PR that is dependent on them. They are not needed. Thanks.

**Review by mrmin123 — CHANGES_REQUESTED** [review]
Requesting removal of the three Util methods you are attempting to add.

Some stylistic guidelines I would like to maintain for consistency and personal preference but am not requiring in this PR (I can clean up in staging after):
* Line breaks in multi-line conditionals after the operator (see second example): https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator
* Hanging indent indentation style (see third 'Yes' example here): https://www.python.org/dev/peps/pep-0008/?#indentation

**Comment by mrmin123** [comment]
Actually, for the sake of a cleaner history, I am going to cherry-pick some changes, clean it up, and then merge to staging. I'll keep this PR open until then.

**Comment by stackhanovets** [comment]
Requesting recheck
And about code makeup: please do it as you wish. Never liked it anyway.

**Comment by mrmin123** [comment]
Closing this PR in favor of #281.

**Comment by stackhanovets** [comment]
Thank you.

I'd like to talk about the code:
```python
                # TODO: recovery_controller.start()
                while not self.regions["lower_right"].exists("next.png"):
                    # TODO: recovery_controller.check()
                    if self.kc_region.exists("combat_nb_fight.png"):
                        self._select_night_battle(self._resolve_night_battle())
                # TODO: recovery_controller.finish()
```

 You've already implemented possible endless loops in `_run_loop_during_battle()` and more, haven't you? That's the reason why recovery will never start during combat phase. To prevent infinite wait time after crashes the `while not Region.exists(Pattern)` construction in fact requires a killswitch (recovery pattern) or countdown timer. You wanted me to clarify my thoughts - here they are::
* `recovery_controller.start()` starts the timer like `Region.wait(Pattern)` does
* `recovery_controller.check()` checks the timer and throw ecxeption after it expiration **and** scans for recovery patterns. Scan events are able to be optimized, e.g each 3rd or 5th iterations etc. `Region.wait(Pattern)` **does not** allow you to implement this easily and cannot be considered flexible.
* `recovery_controller.finish()` releases the timer

The reason - I really hate redundant recovery calls after game lags, especially harsh between 8 and 10 AM JST. 
Again:
* Strategy 1: precise timing for everything. Pros:
  * Ease
  * Predictability
Cons:
  * Lags!
  * Undefinied entry points. E.g. if script crashes at a combat node, it will never go back to port - at least at current release.
* Strategy 2: advanced scanning for crash signatures during waiting time to call recovery when needed, Pros:
  * Fault tolerance by design
  * More relevant to `WYSIWYS` in my opinion
Cons:
  * Overloaded code
  * Possible resources overconsumption
  * Frame freezes are not accounted
  * Also has a problem with entry points

However, this is only a theory. Please leave your remarks or open issue if you want to start a relevant conversation outside of the PR.

**Comment by mrmin123** [comment]
I'm context switching away from this for a couple of hours but will read your comment and respond later tonight, here or in a new issue.

**Comment by mrmin123** [comment]
@stackhanovets I think wrapping the `while not exists` loops in a watcher like that is a perfectly fine way to catch catbombs and crashes if the strategy is to use Sikuli's `observer`s to check for the crashes continuously in the background. That said, it might be perfectly acceptable to insert a simple conditional check within the loops [like in kancolle-auto](https://github.com/mrmin123/kancolle-auto/blob/master/kancolle_auto.sikuli/combat.sikuli/combat.py#L382). I'd also put the relevant code in the `Recovery` module. Either way, I'll take a look at the portions of the code where this is relevant and make a decision based on that.

FYI, some historical context/explanation as to why the `while not exists` loops exists the way they do, and my philosophy on automated recovery and the direction I intend to take it: I've been pretty fortunate and I've only encountered catbombs maybe 4 or 5 times in my entire history of playing kancolle (until this event). When I initially added catbomb support I had to test it by literally turning off the internet to mimic a catbomb. In a perfect world where there are no crashes, the `while not exists` loop perfectly. In kancolle-auto, catbomb recovery was an afterthought so it was finagled in, where I think it worked well? I wouldn't know, I've rarely had to deal with it in the wild. But to be clear, I am not against supporting catbomb/crash recoveries mid-sortie or PvP: it's literally been done before. It's just that it's rarely on my mind so it slips through the cracks (like during the kancolle-auto -> kcauto liftover).

That said, I also do not intend to make kcauto support every possible form of recovery. My thought is that if there is an 'unknown' error (non catbomb/chrome tab crash), the user should deal with it. The script will not and will never forcibly restart the game when it crashes due to new mechanics or lag. For the former, I expect an issue to be opened and the new mechanic to be dealt with systematically. For the latter, I am open to adjusting timings a certain amount, but after a certain point I ask the user to live and deal with it.

Sorry for the info dump. The TLDR is that I'll make a ticket for supporting recovery mid-sortie/PvP and other places where it might be relevant, and general improvements to avoid infinite loops wherever possible, but can't promise that it will never happen, much less say that it's completely avoidable.
